File: chap1/sec1.4/example2Grid.m
l=4;
J=2^l-1;
e=ones(J,1);
h=1/(J+1);
x=0:h:1;
A=1/h^2*spdiags([e -2*e e],[-1 0 1],J,J);
Jc=2^(l-1)-1;                             % coarse grid size
P=sparse(J,Jc);                           % prolongation by interpolation
for j=1:Jc
  P(2*j,j)=1; P(2*j-1,j)=0.5; P(2*j+1,j)=0.5;
end;
R=0.5*P';                                 % restriction
Ac=R*A*P;                                 % coarse matrix by Galerkin 
nu=2;                                     % number of smoothing steps    
rng('default'); u=rand(J,1);              % use same random initial guess
w=2/3;                                    % Jacobi damping parameter  
for n=1:3
  for i=1:nu
    u=u+w*h^2/2*A*u;                      % Jacobi damping step
  end;
  plot(x,[0;u;0],'-');                    % plot before coarse correction
  xlabel('x');ylabel(['error iter = ' num2str(i)])
  axis([0 1 -0.1 1])
  rc=R*(-A*u);                            % compute residual, f=0  
  u=u+P*(Ac\rc);                          % solve coarse problem
  hold on
  plot(x,[0;u;0],'-r');
  legend('before coarse','after coarse')
  hold off
  pause
end;

File: chap1/sec1.4/example2Grid.py
import numpy as np
import scipy.sparse as sp
import scipy.sparse.linalg as spla
import matplotlib.pyplot as plt

# Parameters
l = 4
J = 2**l - 1  # Fine grid size
e = np.ones(J)
h = 1 / (J + 1)
x = np.linspace(0, 1, J + 2)

# Construct the 1D Laplacian matrix on the fine grid
A = 1 / h**2 * sp.diags([e, -2 * e, e], [-1, 0, 1], shape=(J, J))

# Coarse grid parameters
Jc = 2**(l - 1) - 1  # Coarse grid size
P = sp.lil_matrix((J, Jc))  # Prolongation matrix (interpolation)
for j in range(1, Jc + 1):
    P[2 * j - 1, j - 1] = 1
    if 2 * j - 2 >= 0:
        P[2 * j - 2, j - 1] = 0.5
    if 2 * j < J:
        P[2 * j, j - 1] = 0.5
P = P.tocsr()

R = 0.5 * P.T  # Restriction matrix
Ac = R @ A @ P  # Galerkin coarse-grid matrix

# Multigrid parameters
nu = 2  # Number of smoothing steps
np.random.seed(0)
u = np.random.rand(J)  # Random initial guess
w = 2 / 3  # Jacobi damping parameter

# Multigrid V-cycle
plt.figure("example2Grid")
for n in range(3):  # Number of multigrid iterations
    if not plt.fignum_exists("example2Grid"): break

    for i in range(nu):  # Jacobi smoothing steps
        u = u + w * h**2 / 2 * A @ u  # Jacobi damping step

    # Plot before coarse correction
    plt.cla()
    plt.plot(x, np.concatenate(([0], u, [0])), label="Before coarse correction")
    plt.xlabel("x")
    plt.ylabel(f"Error (iteration {n + 1})")
    plt.axis([0, 1, -0.1, 1])
    
    # Compute residual and coarse correction
    rc = R @ (-A @ u)  # Residual (f = 0)
    uc = spla.spsolve(Ac, rc)  # Solve coarse problem
    u = u + P @ uc  # Correct fine-grid solution

    # Plot after coarse correction
    plt.plot(x, np.concatenate(([0], u, [0])), "r", label="After coarse correction")
    plt.legend()
    plt.grid()
    plt.pause(1.5)

plt.show()
